python/pyspark/testing/utils.py

```diff
@@ -209,3 +219,176 @@ def check_error(
+            )
+
+    # special cases: empty datasets, datasets with 0 columns
+    if (df.first() is None and expected.first() is None) or (
+        len(df.columns) == 0 and len(expected.columns) == 0
+    ):
+        return True
+
+    def compare_rows(r1: Row, r2: Row):
+        def compare_vals(val1, val2):
+            if isinstance(val1, list) and isinstance(val2, list):
+                return len(val1) == len(val2) and all(
+                    compare_vals(x, y) for x, y in zip(val1, val2)
+                )
+            elif isinstance(val1, Row) and isinstance(val2, Row):
+                return all(compare_vals(x, y) for x, y in zip(val1, val2))
+            elif isinstance(val1, dict) and isinstance(val2, dict):
+                return (
+                    len(val1.keys()) == len(val2.keys())
+                    and val1.keys() == val2.keys()
+                    and all(compare_vals(val1[k], val2[k]) for k in val1.keys())
+                )
+            elif isinstance(val1, float) and isinstance(val2, float):
+                if abs(val1 - val2) > 1e-5:
+                    return False
+            else:
+                if val1 != val2:
+                    return False
+            return True
+
+        if r1 is None and r2 is None:
+            return True
+        elif r1 is None or r2 is None:
+            return False
+
+        return compare_vals(r1, r2)
+
+    def assert_schema_equal(
+        df_schema: StructType,
+        expected_schema: StructType,
+    ):
+        if df_schema != expected_schema:
+            raise PySparkAssertionError(
+                error_class="DIFFERENT_SCHEMA",
+                message_parameters={"df_schema": df_schema, "expected_schema": expected_schema},
+            )
+
+    def assert_rows_equal(rows1: Row, rows2: Row):
+        zipped = list(zip_longest(rows1, rows2))
+        rows_equal = True
+        error_msg = "Results do not match: "
+        diff_msg = ""
+        diff_rows_cnt = 0
+
+        for r1, r2 in zipped:
+            if not compare_rows(r1, r2):
+                rows_equal = False
+                diff_rows_cnt += 1
+                diff_msg += (
+                    "[df]" + "\n" + str(r1) + "\n\n" + "[expected]" + "\n" + str(r2) + "\n\n"
+                )
+                diff_msg += "********************" + "\n\n"
+
+        if not rows_equal:
+            percent_diff = diff_rows_cnt / len(zipped)
+            error_msg += "( %.5f %% )" % percent_diff
+            error_msg += "\n" + diff_msg
+            raise PySparkAssertionError(
+                error_class="DIFFERENT_ROWS",
+                message_parameters={"error_msg": error_msg},
+            )
+
+    if not check_row_order:
+        try:
+            # rename duplicate columns for sorting
+            renamed_df = df.toDF(*[f"_{i}" for i in range(len(df.columns))])
+            renamed_expected = expected.toDF(*[f"_{i}" for i in range(len(df.columns))])
```

Comment:
`len(df.columns))` should be `len(expected.columns))`?
The number of columns could be different.